iothub/device/tests/ClientPropertyCollectionTests.cs

```diff
@@ -0,0 +1,140 @@
+        }
+
+        [TestMethod]
+        public void ClientPropertyCollection_CanAddSimpleObjectWithComponentAndGetBackWithoutDeviceClient()
+        {
+            var clientProperties = new ClientPropertyCollection
+            {
+                { componentName, propertyName, simplePropertyValue }
+            };
+            clientProperties.TryGetValue<string>(componentName, propertyName, out var outValue);
+            Assert.AreEqual(simplePropertyValue, outValue, $"Property values do not match, expected {simplePropertyValue} but got {outValue}");
+        }
+
+        [TestMethod]
+        public void ClientPropertyCollection_CanUpdateSimpleObjectAndGetBackWithoutDeviceClient()
+        {
+            var clientProperties = new ClientPropertyCollection
+            {
+                { propertyName, simplePropertyValue }
+            };
+            clientProperties.TryGetValue<string>(propertyName, out var outValue);
+            Assert.AreEqual(simplePropertyValue, outValue, $"Property values do not match, expected {simplePropertyValue} but got {outValue}");
+
+            clientProperties.AddOrUpdate(propertyName, updatedPropertyValue);
+            clientProperties.TryGetValue<string>(propertyName, out var outValueChanged);
+            Assert.AreEqual(updatedPropertyValue, outValueChanged, $"Property values do not match, expected {updatedPropertyValue} but got {outValue}");
+        }
+
+        [TestMethod]
+        public void ClientPropertyCollection_CanUpdateSimpleObjectWithComponentAndGetBackWithoutDeviceClient()
+        {
+            var clientProperties = new ClientPropertyCollection
+            {
+                { componentName, propertyName, simplePropertyValue }
+            };
+            clientProperties.TryGetValue<string>(componentName, propertyName, out var outValue);
+            Assert.AreEqual(simplePropertyValue, outValue, $"Property values do not match, expected {simplePropertyValue} but got {outValue}");
+
+            clientProperties.AddOrUpdate(componentName, propertyName, updatedPropertyValue);
+            clientProperties.TryGetValue<string>(componentName, propertyName, out var outValueChanged);
+            Assert.AreEqual(updatedPropertyValue, outValueChanged, $"Property values do not match, expected {updatedPropertyValue} but got {outValue}");
+        }
+
+        [TestMethod]
+        public void ClientPropertyCollection_AddSimpleObjectAgainThrowsException()
+        {
+            var clientProperties = new ClientPropertyCollection
+            {
+                { propertyName, simplePropertyValue }
+            };
+            Assert.ThrowsException<ArgumentException>(() =>
+            {
+                clientProperties.Add(propertyName, simplePropertyValue);
+            });
+        }
+
+        [TestMethod]
+        public void ClientPropertyCollection_AddSimpleObjectWithComponentAgainThrowsException()
```

Comment:
add 2 different properties and both get added?

Reply:
This is handled up a test case. We might need to check to see if we need to change how Add works for component...